src/ngraph/runtime/intelgpu/intelgpu_op_convolution.cpp

```diff
@@ -0,0 +1,265 @@
+/*******************************************************************************
+* Copyright 2017-2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <CPP/custom_gpu_primitive.hpp>
+
+#include "ngraph/runtime/intelgpu/code_writer.hpp"
+#include "ngraph/runtime/intelgpu/intelgpu_layout.hpp"
+#include "ngraph/runtime/intelgpu/intelgpu_op_convolution.hpp"
+#include "ngraph/runtime/intelgpu/intelgpu_op_custom_kernels.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+// this is duplication of the runtime::intelgpu::access_dims
+// needs to be merged but not at the same time as this new code
+static string array_dim(const Shape& dimentions, const string& var = "i", bool is_reversed = false)
+{
+    size_t var_idx = 0;
+    string buffer;
+
+    for (auto const& i : dimentions)
+    {
+        if (is_reversed)
+        {
+            buffer += "[" + to_string(i) + " - " + var + to_string(var_idx) + " - 1]";
+        }
+        else
+        {
+            buffer += "[" + var + to_string(var_idx) + "]";
+        }
+        ++var_idx;
+    }
+
+    if (buffer.empty())
+    { // it means scalar
+        buffer = "[0]";
+    }
+
+    return buffer;
+}
+
+// Padding, Strides and dilation are quite nice explained
+// with animations here https://github.com/vdumoulin/conv_arithmetic
+//
+// batch axes for both input data and output data are 0
+// input channel axes for both input data and filters are 1
+// output channel axes for filters is 0
+// output channel axis for output data is 1
+// Example (Convolution):
+//   data[ 2, 1, 3, 5, 8 ]
```

Comment:
This is a little hard to read. I think if you indented the sample tensors like on lines 63-65 then it would be easier to read.

Reply:
@rkimballn1 I think I didn't understand. Could you please point an example?
Did you mean, better format might be:
```C++
// output channel axis for output data is 1
//
// Example (Convolution):
//       data[              2,              1, 3, 5, 8 ]
//     filter[              2,              1, 2, 2, 3 ]
//     output[              2,              2, 2, 4, 6 ]
// it is like
//       data[          batch,   data_channel, 3, 5, 8 ]
//     filter[ output_channel,   data_channel, 2, 2, 3 ]
//     output[          batch, output_channel, 2, 4, 6 ]
//
```

Reply:
I was thinking something a little easier. No need to line up the numbers
```
// output channel axis for output data is 1
//
// Example (Convolution):
//           data[ 2, 1, 3, 5, 8 ]
//         filter[ 2, 1, 2, 2, 3 ]
//         output[ 2, 2, 2, 4, 6 ]
// it is like
//           data[ batch, data_channel, 3, 5, 8 ]
//         filter[ output_channel, data_channel, 2, 2, 3 ]
//         output[ batch, output_channel, 2, 4, 6 ]
//
```

Reply:
Fixed in PR #1412